helix-term/src/ui/editor.rs

```diff
@@ -276,29 +292,32 @@ impl EditorView {
             )
         };
 
-        // basically just queries the theme color defined in the config
-        let hint = get_scope_of("diagnostic.hint");
-        let info = get_scope_of("diagnostic.info");
-        let warning = get_scope_of("diagnostic.warning");
-        let error = get_scope_of("diagnostic.error");
-        let r#default = get_scope_of("diagnostic"); // this is a bit redundant but should be fine
+        let highlight = match severity {
+            Some(Severity::Error) => get_scope_of("diagnostic.error"),
+            Some(Severity::Warning) => get_scope_of("diagnostic.warning"),
+            Some(Severity::Info) => get_scope_of("diagnostic.info"),
+            Some(Severity::Hint) => get_scope_of("diagnostic.hint"),
+            None => get_scope_of("diagnostic"),
+        };
 
-        doc.diagnostics()
-            .iter()
-            .map(|diagnostic| {
-                let diagnostic_scope = match diagnostic.severity {
-                    Some(Severity::Info) => info,
-                    Some(Severity::Hint) => hint,
-                    Some(Severity::Warning) => warning,
-                    Some(Severity::Error) => error,
-                    _ => r#default,
-                };
-                (
-                    diagnostic_scope,
-                    diagnostic.range.start..diagnostic.range.end,
-                )
-            })
-            .collect()
+        let diagnostic_ranges = doc.diagnostics().iter().filter_map(move |diagnostic| {
```

Comment:
We end up filtering `doc.diagnostics()` for each severity, right? Could we iterate over it once and separate out the diagnostics into a `HashMap<Severity, SmallVec<Range<usize>>` to save some work?